apps/rush-lib/src/cli/utilities/pnpm/PnpmLinkManager.ts

```diff
@@ -0,0 +1,210 @@
+import * as os from 'os';
+import * as path from 'path';
+import * as semver from 'semver';
+import encodeRegistry = require('encode-registry');
+
+import { JsonFile } from '@microsoft/node-core-library';
+
+import {
+  BaseLinkManager,
+  SymlinkKind
+} from '../base/BaseLinkManager';
+import Utilities from '../../../utilities/Utilities';
+import { BasePackage } from '../base/BasePackage';
+import { RushConstants } from '../../../RushConstants';
+import { IRushLinkJson } from '../../../data/RushConfiguration';
+import RushConfigurationProject from '../../../data/RushConfigurationProject';
+
+export class PnpmLinkManager extends BaseLinkManager {
+  protected _linkProjects(): Promise<void> {
+    return new Promise<void>((resolve: () => void, reject: (reason: Error) => void): void => {
+      // go ahead and find out the registry we are using, we need this to locate the pnpm store folder
+
+      // EXAMPLE: https://onedrive.pkgs.visualstudio.com/_packaging/odsp-npm/npm/registry
+      const registryUrl: string = Utilities.executeCommandAndCaptureOutput(
+        this._rushConfiguration.packageManagerToolFilename,
+        ['config', 'get', 'registry'],
+        this._rushConfiguration.commonTempFolder,
+        process.env);
+
+      // EXAMPLE: onedrive.pkgs.visualstudio.com
+      const encodedRegistry: string = encodeRegistry(registryUrl);
+
+      // EXAMPLE: c:/src/foo/common/temp/node_modules/.onedrive.pkgs.visualstudio.com
+      const registryPath: string = path.join(
+        this._rushConfiguration.commonTempFolder,
+        RushConstants.nodeModulesFolderName,
+        `.${encodedRegistry}`
+      );
+
+      try {
+        const rushLinkJson: IRushLinkJson = { localLinks: {} };
+
+        for (const rushProject of this._rushConfiguration.projects) {
+          console.log(os.EOL + 'LINKING: ' + rushProject.packageName);
+          this._linkProject(registryPath, rushProject, rushLinkJson);
+        }
+
+        console.log(`Writing "${this._rushConfiguration.rushLinkJsonFilename}"`);
+        JsonFile.save(rushLinkJson, this._rushConfiguration.rushLinkJsonFilename);
+
+        resolve();
+      } catch (error) {
+        reject(error);
+      }
+    });
+  }
+
+  /**
+   * This is called once for each local project from Rush.json.
+   * @param project             The local project that we will create symlinks for
+   * @param rushLinkJson        The common/temp/rush-link.json output file
+   */
+  private _linkProject(
+    pnpmStorePath: string,
+    project: RushConfigurationProject,
+    rushLinkJson: IRushLinkJson): void {
+
+    // PNPM creates a directed acyclic graph, rather than a tree.
+    // Thus, imagine "sp-http" and "sp-core-library" both depend on "request"@1.2.3,
+    // and "request"@1.2.3 depends on "http@*", then both "sp-http" and "sp-core-library"
+    // must have the exact same version of "http"
+    // thus, we only need to link the direct dependencies of this project to the
+    // common folder
```

Comment:
With NPM, it's possible for two different projects to have dependencies on the same version of the same library, but end up with different implementations of that library, if the library is installed twice and with different secondary dependencies.  The NpmLinkManager recursively links dependency folders to try to honor this.  Since PNPM always uses the same physical folder to represent a given version of a library, we only need to link directly to that folder, and it will have a consistent set of secondary dependencies.

Reply:
What happens in this situation?

```
Project1:
  Library@1.2.3
  React@1

Project 2:
  Library@1.2.3
  React@2

Library@1.2.3:
  React@*
```

---
In reply to: [152452153](https://github.com/Microsoft/web-build-tools/pull/449#discussion_r152452153) [](ancestors = 152452153)

Reply:
Looking at that situation now

---
In reply to: [152452232](https://github.com/Microsoft/web-build-tools/pull/449#discussion_r152452232) [](ancestors = 152452232,152452153)